password-cli/app/Main.hs

```diff
@@ -0,0 +1,141 @@
+        )
+
+commandHelp :: Parser (IO ())
+commandHelp =
+  go
+    <$> optional (argument str (metavar "COMMAND"))
+    <**> helper
+  where
+    go mCmd =
+      let args = maybe id (:) mCmd ["-h"]
+       in void $ handleParseResult $ execParserPure defaultPrefs cliOpts args
+
+commandArgon2 :: Parser (IO ())
+commandArgon2 =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash Argon2.hashPassword
+    commandCheck hash = runCheck $ \password -> Argon2.checkPassword password $ PasswordHash hash
+
+commandBcrypt :: Parser (IO ())
+commandBcrypt =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash Bcrypt.hashPassword
+    commandCheck hash = runCheck $ \password -> Bcrypt.checkPassword password $ PasswordHash hash
+
+commandPBKDF2 :: Parser (IO ())
+commandPBKDF2 =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash PBKDF2.hashPassword
+    commandCheck hash = runCheck $ \password -> PBKDF2.checkPassword password $ PasswordHash hash
+
+commandScrypt :: Parser (IO ())
+commandScrypt =
+  subparser
+    ( command
+        "hash"
+        ( info
+            (pure commandHash)
+            (progDesc "hash password (via STDIN)")
+        )
+        <> command
+          "check"
+          ( info
+              (commandCheck <$> option str (metavar "HASH" <> long "hash") <**> helper)
+              (progDesc "check hashed password (via STDIN)")
+          )
+    )
+    <**> helper
+  where
+    commandHash = runHash Scrypt.hashPassword
+    commandCheck hash = runCheck $ \password -> Scrypt.checkPassword password $ PasswordHash hash
+
+runHash :: (Password -> IO (PasswordHash a)) -> IO ()
+runHash f = do
+  password <- mkPassword <$> T.hGetLine stdin
+  hash <- f password
+  T.putStr $ unPasswordHash hash
+
+runCheck :: (Password -> PasswordCheck) -> IO ()
+runCheck f = do
+  password <- mkPassword <$> T.hGetLine stdin
```

Comment:
@cdepillabout Do you remember something about being advised to not use `Data.Text.IO`? I've a weak recollection of having read that somewhere :thinking:

Reply:
I think the idea is that `Data.Text.IO` (and similar functions from `base`) use the user's encoding when reading in values:

https://www.snoyman.com/blog/2016/12/beware-of-readfile/

I'm not sure of the exact consequences here.

Reply:
Ah, I think it was `readFile` indeed. I think I read it here:

https://www.snoyman.com/blog/2020/10/haskell-bad-parts-1/

Which is also snoyman, but later :sweat_smile: